File: android/src/main/java/com/nitro/thermalprinter/ImageCache.kt
package com.nitro.thermalprinter

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL

/**
 * LRU Image cache for thermal printer images.
 * Caches up to maxSize images, evicting least recently used when full.
 * Uses URL as key for caching.
 */
class ImageCache(private val maxSize: Int = 10) {

    private val cache = object : LinkedHashMap<String, Bitmap>(maxSize, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Bitmap>?): Boolean {
            return size > maxSize
        }
    }

    /**
     * Cache an image from URL
     * @param url Image URL to download and cache
     * @param key Unique key to reference the cached image
     */
    suspend fun cacheFromUrl(url: String, key: String) {
        withContext(Dispatchers.IO) {
            val bitmap = downloadBitmap(url)
            bitmap?.let {
                synchronized(cache) {
                    cache[key] = it
                }
            } ?: throw IllegalArgumentException("Failed to download image from: $url")
        }
    }

    /**
     * Cache a bitmap directly
     * @param bitmap Bitmap to cache
     * @param key Unique key to reference the cached image
     */
    fun cacheBitmap(bitmap: Bitmap, key: String) {
        synchronized(cache) {
            cache[key] = bitmap
        }
    }

    /**
     * Get a cached image by key
     * @param key Key used when caching
     * @returns Cached bitmap or null if not found
     */
    fun get(key: String): Bitmap? {
        return synchronized(cache) {
            cache[key]
        }
    }

    /**
     * Check if an image is cached
     * @param key Key to check
     * @returns true if image is cached
     */
    fun contains(key: String): Boolean {
        return synchronized(cache) {
            cache.containsKey(key)
        }
    }

    /**
     * Remove a cached image
     * @param key Key of image to remove
     */
    fun remove(key: String) {
        synchronized(cache) {
            cache.remove(key)?.let { bitmap ->
                if (!bitmap.isRecycled) bitmap.recycle()
            }
        }
    }

    /**
     * Clear all cached images
     */
    fun clear() {
        synchronized(cache) {
            cache.values.forEach { bitmap ->
                if (!bitmap.isRecycled) bitmap.recycle()
            }
            cache.clear()
        }
    }

    /**
     * Get current cache size
     */
    fun size(): Int {
        return synchronized(cache) {
            cache.size
        }
    }

    /**
     * Download a bitmap from URL
     */
    private fun downloadBitmap(src: String): Bitmap? {
        return try {
            val url = URL(src)
            val connection = url.openConnection() as HttpURLConnection
            connection.doInput = true
            connection.connectTimeout = 10000
            connection.readTimeout = 10000
            connection.connect()

            val inputStream = connection.inputStream
            val bitmap = BitmapFactory.decodeStream(inputStream)
            inputStream.close()
            connection.disconnect()

            bitmap
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * Download and decode a bitmap from URL (public version for direct use)
     */
    suspend fun downloadBitmapAsync(url: String): Bitmap? {
        return withContext(Dispatchers.IO) {
            downloadBitmap(url)
        }
    }
}

File: ios/Sources/ImageCache.swift
import UIKit

/// LRU Image cache for thermal printer images.
/// Caches up to maxSize images, evicting least recently used when full.
/// Uses URL as key for caching.
internal class ImageCache {
    private let maxSize: Int
    private var cache: [String: UIImage] = [:]
    private var accessOrder: [String] = []
    private let lock = NSLock()

    public init(maxSize: Int = 10) {
        self.maxSize = maxSize
    }

    /// Cache an image from URL
    public func cacheFromUrl(_ url: String, key: String) async throws {
        guard let imageUrl = URL(string: url) else {
            throw ImageCacheError.invalidUrl
        }

        let (data, _) = try await URLSession.shared.data(from: imageUrl)

        guard let image = UIImage(data: data) else {
            throw ImageCacheError.invalidImageData
        }

        lock.lock()
        defer { lock.unlock() }

        // Evict oldest if at capacity
        if cache.count >= maxSize, let eldest = accessOrder.first {
            cache.removeValue(forKey: eldest)
            accessOrder.removeFirst()
        }

        cache[key] = image
        accessOrder.append(key)
    }

    /// Cache an image directly
    public func cacheImage(_ image: UIImage, key: String) {
        lock.lock()
        defer { lock.unlock() }

        // Evict oldest if at capacity
        if cache.count >= maxSize && !cache.keys.contains(key), let eldest = accessOrder.first {
            cache.removeValue(forKey: eldest)
            accessOrder.removeFirst()
        }

        // Update access order if key exists
        if let index = accessOrder.firstIndex(of: key) {
            accessOrder.remove(at: index)
        }

        cache[key] = image
        accessOrder.append(key)
    }

    /// Get a cached image by key
    public func get(key: String) -> UIImage? {
        lock.lock()
        defer { lock.unlock() }

        guard let image = cache[key] else { return nil }

        // Update access order (LRU)
        if let index = accessOrder.firstIndex(of: key) {
            accessOrder.remove(at: index)
            accessOrder.append(key)
        }

        return image
    }

    /// Check if an image is cached
    public func contains(key: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return cache.keys.contains(key)
    }

    /// Remove a cached image
    public func remove(key: String) {
        lock.lock()
        defer { lock.unlock() }

        cache.removeValue(forKey: key)
        if let index = accessOrder.firstIndex(of: key) {
            accessOrder.remove(at: index)
        }
    }

    /// Clear all cached images
    public func clear() {
        lock.lock()
        defer { lock.unlock() }

        cache.removeAll()
        accessOrder.removeAll()
    }

    /// Get current cache size
    public func size() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return cache.count
    }

    /// Download an image from URL
    public func downloadImage(from url: String) async throws -> UIImage {
        guard let imageUrl = URL(string: url) else {
            throw ImageCacheError.invalidUrl
        }

        let (data, _) = try await URLSession.shared.data(from: imageUrl)

        guard let image = UIImage(data: data) else {
            throw ImageCacheError.invalidImageData
        }

        return image
    }
}

/// Errors for ImageCache
internal enum ImageCacheError: Error, LocalizedError {
    case invalidUrl
    case invalidImageData
    case downloadFailed

    public var errorDescription: String? {
        switch self {
        case .invalidUrl:
            return "Invalid image URL"
        case .invalidImageData:
            return "Invalid image data"
        case .downloadFailed:
            return "Failed to download image"
        }
    }
}
